File: pid_control/script/pid_controller.py
#!/usr/bin/env python

import rospy
import numpy as np
from sensor_msgs.msg import Imu
from geometry_msgs.msg import Twist
from geometry_msgs.msg import Vector3


def clamp(num, min_value, max_value):
   return max(min(num, max_value), min_value)

def callback(message):
    angle = message.orientation.x
    print (angle+0.5)
    res = Twist()
    #res.linear.x = 1
    res.angular.z = -clamp(pwm(angle+0.5),-20,+20)
    publisher = rospy.Publisher('/cmd_vel',Twist, queue_size=10)
    publisher.publish(res)
    loop_rate.sleep()

def pwm(angle, kp = 1000, ki = 1 , kd = 10):
    global i
    global oldi
    global oldp
    
    p = angle * kp
    i = i + (p * 0.05)
    i = i + ((i - oldi) * 2)
    i = clamp(i,-250,+250)
    d = p - oldp
    res = p + i * ki + d * kd

    oldp = p
    oldi = i

    return res

if __name__ == "__main__":
    oldi = 0
    oldp = 0
    i = 0

    rospy.init_node('controller' , anonymous=True)
    loop_rate = rospy.Rate(100)
    while not rospy.is_shutdown():
        rospy.Subscriber('imu', Imu , callback)
        rospy.spin()

File: pid_control/pid_controller.cpp
#include "ros/ros.h"
#include "ros/console.h"
#include "sensor_msgs/Imu.h"
#include "geometry_msgs/Twist.h"


double clamp(double num, double min_value, double max_value)
{
    return std::max(std::min(num, max_value), min_value);
}

double map(double x,double max_value, double min_value , double min , double max)
{
    return min + (min - max) * (x - min_value) / (min_value - max_value);
}

class myController
{
    public:

        double i = 0 ;
        double oldp = 0;
        double oldi = 0;
        double p = 0;
        double d = 0;

        myController(void)
        {
            ros::NodeHandle nh;
            ros::Rate loop_rate(100);

            imu_sub = nh.subscribe("imu",5,&myController::call_back,this);
            vel_pub = nh.advertise<geometry_msgs::Twist>("cmd_vel",10,false);
        }

        void call_back(const sensor_msgs::Imu::ConstPtr& msg)
        {
            double angle = msg->orientation.x + 0.5;
            //ROS_DEBUG("angle = %", angle);
            std::cout << angle << std::endl;
            geometry_msgs::Twist res;
            res.angular.z = -map(pwm(angle+0.5),-250,250,-15,+15);
            std::cout << pwm(angle+0.5) << std::endl;
            vel_pub.publish(res);
        }

    private:
        ros::Subscriber imu_sub;
        ros::Publisher vel_pub;

        double pwm(float angle, float kp = 1000, float ki = 1 , float kd = 10)
        {
            p = angle * kp;
            i = i + (p * 0.05);
            i = i + ((i - oldi) * 2);
            i = clamp (i , -250 , +250);
            d = p - oldp;

            oldp = p;
            oldi = i;

            return clamp(p + i * ki + d * kd,-250,+250);
        }

};

int main(int argc, char **argv)
{
    ros::init(argc,argv,"controller");
    
    myController ctr;

    ros::spin();

}